scripts/dataset_sorting.py

```diff
@@ -0,0 +1,183 @@
+        mzs = np.concatenate(mzs_list)
+        by_mz = np.argsort(mzs)
+        sp_mz_int_buf = np.array([np.concatenate(sp_inds_list)[by_mz],
+                                  mzs[by_mz],
+                                  np.concatenate(ints_list)[by_mz]], dtype).T
+        return sp_mz_int_buf
+
+    curr_sp_i = 0
+    sp_inds_list, mzs_list, ints_list = [], [], []
+
+    estimated_size_mb = 0
+    for x, y in coordinates:
+        mzs_, ints_ = imzml_parser.getspectrum(curr_sp_i)
+        mzs_, ints_ = map(np.array, [mzs_, ints_])
+        sp_idx = sp_id_to_idx[curr_sp_i]
+        sp_inds_list.append(np.ones_like(mzs_) * sp_idx)
+        mzs_list.append(mzs_)
+        ints_list.append(ints_)
+        estimated_size_mb += 2 * mzs_.nbytes + ints_.nbytes
+        curr_sp_i += 1
+        if estimated_size_mb > max_size:
+            yield sort_dataset_chunk(sp_inds_list, mzs_list, ints_list)
+            sp_inds_list, mzs_list, ints_list = [], [], []
+            estimated_size_mb = 0
+
+    if len(sp_inds_list) > 0:
+        yield sort_dataset_chunk(sp_inds_list, mzs_list, ints_list)
+
+
+def define_ds_segments(imzml_parser, sp_n, ds_segm_size_mb=5, sample_sp_n=1000):
+    print('Defining segments bounds...', end=' ', flush=True)
+    t = time()
+
+    sample_ratio = sample_sp_n / sp_n
+
+    def spectra_sample_gen(imzml_parser, sample_ratio):
+        sp_n = len(imzml_parser.coordinates)
+        sample_size = int(sp_n * sample_ratio)
+        sample_sp_inds = np.random.choice(np.arange(sp_n), sample_size)
+        for sp_idx in sample_sp_inds:
+            mzs, ints = imzml_parser.getspectrum(sp_idx)
+            yield sp_idx, mzs, ints
+
+    spectra_sample = list(spectra_sample_gen(imzml_parser, sample_ratio=sample_ratio))
+
+    spectra_mzs = np.array([mz for sp_id, mzs, ints in spectra_sample for mz in mzs])
+    total_n_mz = spectra_mzs.shape[0] / sample_ratio
+
+    float_prec = 4 if imzml_parser.mzPrecision == 'f' else 8
+    segm_arr_columns = 3
+    segm_n = segm_arr_columns * (total_n_mz * float_prec) // (ds_segm_size_mb * 2 ** 20)
+    segm_n = max(1, int(segm_n))
+
+    segm_bounds_q = [i * 1 / segm_n for i in range(0, segm_n + 1)]
+    segm_lower_bounds = [np.quantile(spectra_mzs, q) for q in segm_bounds_q]
+    ds_segments_bounds = np.array(list(zip(segm_lower_bounds[:-1], segm_lower_bounds[1:])))
+
+    max_mz_value = 10 ** 5
+    ds_segments_bounds[0, 0] = 0
+    ds_segments_bounds[-1, 1] = max_mz_value
+
+    print('DONE {:.2f} sec'.format(time() - t))
+    print(' * segments number:', segm_n)
+    return ds_segments_bounds
+
+
+def segment_spectra_chunk(sp_mz_int_buf, ds_segments_bounds, ds_segments_path):
+    for segm_i, (l, r) in enumerate(ds_segments_bounds):
+        segm_start, segm_end = np.searchsorted(sp_mz_int_buf[:, 1], (l, r))
+        pd.to_msgpack(f'{ds_segments_path}/ds_segm_{segm_i:04}.msgpack',
+                      sp_mz_int_buf[segm_start:segm_end],
+                      append=True)
```

Comment:
@LachlanStuart actually, the segments are already in a dataframe format

Reply:
No, they're Numpy ndarrays. You can read one back from the disk and check, e.g. `pd.read_msgpack('./segms/ds_segm_0000.msgpack')`.

`pd.to_msgpack` and `pd.read_msgpack` are just wrappers around a serialization library for convenience. Those functions don't actually require the data to be DataFrames.